### PR TITLE
Add GNS Logging

### PR DESCRIPTION
**Pull request by gkielian**
Adding results from NeurIPS 2024 for GNS monitoring.

This appears to double the time per inference, adding a new field.

It doesn't appear at the moment to make any changes to the output.

It might be possible to dynamically modify the batch size manually reading this data.

Relates to the following issue -- but probably should create a new issue dedicated for GNS:

- https://github.com/ReaLLMASIC/nanoGPT/issues/350
- https://arxiv.org/abs/2411.00999

Currently this only supports single dataset training.

The authors showed that we should aim to create a schedule of batch size (similar to learning rate schedule) and but that should be changed linearly with the number of total tokens processed.

By doing so they were able to improve reach the same perplexity with 18% less training tokens, and thus with the same number of training tokens, would have a stronger performance per parameter.

### CODE REVIEW

**Review by klei22 — APPROVED** [review]
Looks good
